### PR TITLE
Fluent Java Interop & Method Dispatch Refactor

### PR DESCRIPTION
**Pull request by kustosz**
### Pull Request Description
Removes currying for polyglot calls while also removing the need for vectored arguments and unifying field/method access patterns. Also pushes the method dispatch logic to a truffle library, so that our resolution logic can be more object-oriented and less "let me try to handle every single type of value in one node".

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: distribution/std-lib/Base/src/Data/Text/Extensions.enso [hunk 1]
```diff
@@ -37,15 +37,15 @@ Text.length =
...
 Text.each function =
-    iterator = BreakIterator.getCharacterInstance []
-    iterator.setText [this]
+    iterator = BreakIterator.getCharacterInstance
+    iterator.setText this
 
-    fst = iterator.first []
-    nxt = iterator.next []
+    fst = iterator.first
+    nxt = iterator.next
```
Comment (iamrecursion) [inline distribution/std-lib/Base/src/Data/Text/Extensions.enso:44]: Surely we can use `next` here now, given that we no longer have UFCS! Other locations too.

File: distribution/std-lib/Base/src/Polyglot/Java.enso [hunk 2]
```diff
@@ -4,5 +4,5 @@ import Builtins
 ## Checks whether an object is an instance of a given class.
 Builtins.Java.is_instance : Any -> Any -> Boolean
 Builtins.Java.is_instance object class =
-    class_object = Polyglot.get_member class "class"
-    class_object.isInstance [object]
+    class_object = class.class
```
Comment (iamrecursion) [inline distribution/std-lib/Base/src/Polyglot/Java.enso:7]: :'D

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/IndirectInvokeMethodNode.java [hunk 3]
```diff
@@ -53,204 +56,36 @@ public abstract Stateful execute(
...
+          function,
+          frame,
+          state,
+          arguments,
+          schema,
+          defaultsExecutionMode,
+          argumentsExecutionMode,
+          isTail);
+    } catch (MethodDispatchLibrary.NoSuchMethodException e) {
+      CompilerDirectives.transferToInterpreter();
```
Comment (iamrecursion) [inline engine/runtime/src/main/java/org/enso/interpreter/node/callable/IndirectInvokeMethodNode.java]: We really need a consistent rule for whether we use this or not when throwing.

File: test/Benchmarks/src/Collections.enso [hunk 4]
```diff
@@ -19,8 +19,8 @@ sum_list_meta list =
 sum_recur n = if n == 0 then 0 else 1 + here.sum_recur n-1
 
 build_map size =
-    rand = Random.new [].to_array
-    0.up_to size . fold Map.empty (m -> i -> m.insert (rand.nextInt [10000]) i)
+    rand = Random.new
+    0.up_to size . fold Map.empty (m -> i -> m.insert (rand.nextInt 10000) i)
```
Comment (iamrecursion) [inline test/Benchmarks/src/Collections.enso:23]: Did you fix the other benchmarks? They use polyglot a lot.

**Review by iamrecursion — APPROVED** [review]

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/IndirectInvokeMethodNode.java (hunk 3 above)
Comment (kustosz) [reply engine/runtime/src/main/java/org/enso/interpreter/node/callable/IndirectInvokeMethodNode.java]: Fair. Let's stick with `don't use` for new code.

File: distribution/std-lib/Base/src/Data/Text/Extensions.enso (hunk 1 above)
Comment (kustosz) [reply distribution/std-lib/Base/src/Data/Text/Extensions.enso:44]: We sure can. But this PR had too much boy-scouting in other places already, it's just too much for one person :P

**Pull request merged**